src/charm.py

```diff
@@ -289,6 +291,30 @@ def _install_cni_binaries(self):
 
         log.info(f"Extracted 'cni-plugins' to {unpack_path}")
 
+    def _on_upgrade_action(self, event):
+        """Handle the upgrade action."""
+        with status.context(self.unit):
+            channel = self.model.config.get("channel")
+            log.info(f"Starting the upgrade of Kubernetes snaps to '{channel}' channel.")
+
+            try:
+                kubernetes_snaps.install(channel=channel, control_plane=True, upgrade=True)
+            except (CalledProcessError, Exception) as e:
+                if isinstance(e, CalledProcessError):
+                    error_message = f"Upgrade failed with a process error. stdout: {e.stdout}, stderr: {e.stderr}"
+                else:
+                    error_message = f"An unexpected error occurred during the upgrade: {e}"
+
+                log.exception(error_message)
+                status.add("Snap upgrade failed. Check action results for more information.")
+                event.fail(error_message)
+            else:
+                result_message = (
+                    f"Successfully upgraded Kubernetes snaps to the '{channel}' channel."
+                )
+                log.info(result_message)
+                event.set_results({"result": result_message})
+
```

Comment:
this is same handler in both charms except for the `control_plane=True` isn't it?  

any chance you'd like to bury the handler in the `kubernetes_snaps` library?

Reply:
the `event` should have a link to the framework -- and the framework should have some metadata.  and that metadata should know the `charm-name` (not application name)

Reply:
or -- you could write a SMALL handler in each charm that passes the right flag to the library handler

Reply:
Is it possible to retrieve the current unit and the `channel` config value from the event? I considered this approach, but then the afterthought made me reconsider...

Reply:
so the event has in it the [`framework`](https://ops.readthedocs.io/en/latest/#ops.Framework) which exposes:
* `meta` -- metadata about the charm
* `model` 
   * model has the `config` property
   * model has the `unit` property